guardian/prosemirror-editor
# Get minimum ProseMirror editor working

Co-authored with @samanthagottlieb

## What does this change?

This gets a basic ProseMirror editor working. The aim is to create a general purpose editor that satisfied the requirements of:
- Primarily:
  - [tagmanager](https://github.com/guardian/tagmanager/blob/c751b1f0da6d61a6e7bd3c6cb43ceddcd743fd9c/public/components/utils/ReactScribe.react.js#L8)
  - [atom-workshop](https://github.com/guardian/atom-workshop/blob/811f99c5ea04d6b55cbddb5dc5c08d0165f23fa8/public/js/components/FormFields/FormFieldScribeEditor.js#L11)
  - [targeting](https://github.com/guardian/targeting/blob/main/public/components/utils/ReactScribe.js)
- Secondly:
  - media-atom-maker

media-atom-maker is a lower priority because it's already been moved to ProseMirror from Scribe (so there's no Scribe-based vulnerability), but we want to make this package compatible with media-atom-maker's requirements so it can be ported over in the future, so that the editors in the four repositories can all be managed via this one package.

This branch has been tested locally in tagmanager via `yalc`. The package has not yet been deployed to `npm`.

Things this PR doesn't have, that we will want to add in future PRs:
- A means of testing changes from this package itself, for example a demo-app that includes the React editor component. Currently this package only builds a component to be used by other packages.
- A means of deploying to an `npm` package.

Currently, vanilla css styles are included in the built `dist` and need to be imported in the target package. Is this a good setup? (A google chat thread led to recommendations against `emotion` and I wanted to keep this package as un-opinionated as possible).

## How to test

1. Publish this package locally using `yarn yalc` from the root of the project
2. Use [this branch](https://github.com/guardian/tagmanager/pull/500) of `tagmanager` locally.
3. Add that package to tagmanager using `yalc add @guardian/prosemirror-editor`
4. Run tagmanager
5. In tagmanager, go to 'Create a new tag' and pick an option from the dropdown. 
6. Click on an existing tag from the home page table and inspect the rich text editor element of the Description/Profile field. Do the classes include 'prosemirror'? (They should).
7. Open the network tab, clear the requests. Then edit the tag's rich text, and save the changes. A POST request should go to the url similar to `https://tagmanager.local.dev-gutools.co.uk/api/tag/71301`. Does the description sent in that request look valid? Is the bold text serialised as `<b>` tags and the italic text as `<i>` tags? (It should). Are new paragraphs serialised as new `<p>` tags? (They should).
